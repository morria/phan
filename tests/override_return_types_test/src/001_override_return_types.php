<?php

class Animal {}

class Cat extends Animal {}

class Dog extends Animal {}

/* @phan-suppress-next-line PhanUnreferencedClass */
class Vegetable {}

/* @phan-suppress-next-line PhanUnreferencedClass */
class OverrideReturnTypesTest {

    public function getAnimal(): Animal {
        return random_int(0, 10) > 5 ? new Cat : new Dog;
    }

    /**
     * This method has an incorrect phpdoc @return annotation.
     * @return Vegetable
     */
    public function getAnimal2() {
        return new Cat;
    }

    /* @phan-suppress-next-line PhanUnreferencedPublicMethod */
    public function run(): void {
        $this->plant($this->getAnimal());
        $this->plant($this->getAnimal2());
    }

    public function plant(Vegetable $v): void {
        var_dump($v);
    }

}
