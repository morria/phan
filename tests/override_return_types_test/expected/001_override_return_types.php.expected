src/001_override_return_types.php:24 PhanTypeMismatchReturnProbablyReal Returning new Cat() of type \Cat but getAnimal2() is declared to return \Vegetable (no real type) (the inferred real return type has nothing in common with the declared phpdoc return type)
src/001_override_return_types.php:29 PhanTypeMismatchArgumentReal Argument 1 ($v) is $this->getAnimal() of type \Animal|\Cat|\Dog but \OverrideReturnTypesTest::plant() takes \Vegetable defined at src/001_override_return_types.php:33
src/001_override_return_types.php:30 PhanPartialTypeMismatchArgument Argument 1 ($v) is $this->getAnimal2() of type \Cat|\Vegetable but \OverrideReturnTypesTest::plant() takes \Vegetable (\Animal|\Cat is incompatible) defined at src/001_override_return_types.php:33
